src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/FindAllReferencesHandler.cs

```diff
@@ -4,22 +4,43 @@
 using System;
 using System.Collections.Generic;
 using System.Composition;
+using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Client;
 using Microsoft.VisualStudio.LanguageServer.ContainedLanguage;
 using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Threading;
+using Newtonsoft.Json.Linq;
 
 namespace Microsoft.VisualStudio.LanguageServerClient.Razor.HtmlCSharp
 {
     [Shared]
     [ExportLspMethod(Methods.TextDocumentReferencesName)]
-    internal class FindAllReferencesHandler : IRequestHandler<ReferenceParams, VSReferenceItem[]>
+    internal class FindAllReferencesHandler : IRequestHandler<ReferenceParams, VSReferenceItem[]>, IDisposable
     {
         private readonly LSPRequestInvoker _requestInvoker;
         private readonly LSPDocumentManager _documentManager;
         private readonly LSPProjectionProvider _projectionProvider;
         private readonly LSPDocumentMappingProvider _documentMappingProvider;
 
+        private readonly TimeSpan WaitForProgressCompletionTimeout = new TimeSpan(0, 0, 15);
+
+        // Roslyn sends Progress Notifications every 0.5s *only* if results have been found.
+        // Consequently, at ~ time > 0.5s ~ after the last notification, we don't know whether Roslyn is
+        // done searching for results, or just hasn't found any additional results yet.
+        // 
+        // https://github.com/dotnet/roslyn/blob/1462b1e9c78a7efe338a6859ee5f88fa07d4adc8/src/Features/LanguageServer/Protocol/CustomProtocol/FindUsagesLSPContext.cs#L52-L55
+        // 
+        // To work around this, we wait for up to 3s since the last notification before timing out.
+        // 
+        private readonly TimeSpan WaitForProgressNotificationTimeout = new TimeSpan(0, 0, 0, 3);
+
+        private IProgress<object> ActiveRequest { get; set; }
```

Comment:
Because there's only 1 of these what happens when you do multiple FAR requests on different locations in a row when they're all going simultaneously? Is that a scenario? I ask because technically there's a "KeepResults" option in the FAR window and i'm not sure how that operates when a previous request hasn't finished yet.

Reply:
With the addition of these pieces it makes this class stateful. I'd suggest not doing that. Instead I'd recommend building a request manager private class in this file that manages FAR requests as they come in and handles updating their various progresses.

Reply:
There's a bit of context here which I should have better documented. 

_Overall, is this implementation, as it stands, stateful?_
Yes, but... this should change once VS LSP fixes a [bug](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1081157/) with regards to the serialization of the `IProgress` object. Right now, the `ReferenceParams.PartialResultToken` we send to the VS LSP is serialized to `{}`. This has the important consequence of losing the token (id) of the FAR request. Once this is fixed, and the ID is preserved, we should have the [token for a given request in the progress notification](https://github.com/dotnet/aspnetcore-tooling/pull/2112/files#diff-4d367dc180f9d8b2a85502e662441dc9R94-R95).

With this token, we should be able to maintain an internal hash-map of <token, IProgress>, which should allow us to better handle multiple requests and remove the current statefulness.


> Because there's only 1 of these what happens when you do multiple FAR requests on different locations in a row when they're all going simultaneously? Is that a scenario? 

Right now, the last request will be kept / acted on. We reset the "state" whenever a [new FAR request](https://github.com/dotnet/aspnetcore-tooling/pull/2112/files#diff-4d367dc180f9d8b2a85502e662441dc9R163-R164) is received. This reset ensures we no longer report any more results from the previous request as we reset the `IProgress` object used for reporting.

This could still cause issues if the previous request gets subsequent progress notifications, however this will be resolved once the token becomes available (as we'll be able to identify the original FAR request of a given progress notification).

> I ask because technically there's a "KeepResults" option in the FAR window and i'm not sure how that operates when a previous request hasn't finished yet.

Haven't tried Keep Results, let me get back to you on that.

> With the addition of these pieces it makes this class stateful. I'd suggest not doing that. Instead I'd recommend building a request manager private class in this file that manages FAR requests as they come in and handles updating their various progresses.

Given the context I've now provided, do you still feel the same way regarding an additional request manager class?

Reply:
Ahh if most of this statefulness is temporary that definitely changes my perception. Is there any reason we can't just define our own IProgress<T> type that wraps /inherits from theirs which happens to serialize correctly?

Reply:
> Is there any reason we can't just define our own IProgress type that wraps /inherits from theirs which happens to serialize correctly?

Yeah, I got it working with a very basic wrapper which essentially made the `PartialResultToken` a string instead of `IProgress` thereby working around the bug. I could write up the token handling against that. Just focused on the rest first, and was advised that this bug should be fixed this week so figured might as well wait to see if there are any other resultant changes. I'll do the token handling now then.